components/brave_wallet_ui/common/hardware/ledgerjs/ledger-messaging-transport.ts

```diff
@@ -0,0 +1,96 @@
+/* Copyright (c) 2022 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+import {
+  LedgerErrorsCodes,
+  LedgerFrameCommand
+} from './ledger-messages'
+
+// LedgerMessagingTransport is a generic bi-directional messaging utility for
+// Window objects. It supports supports both (1) the sending of messages via postMessage
+// to a window object at a targetUrl and subscribing of responses, and (2)
+// the definition of handlers to be run when a different LedgerMessagingTransport
+// instance at a different Window sends messages to it.
+export abstract class LedgerMessagingTransport {
+  protected targetWindow: Window
+  protected targetUrl: string
+  protected handlers: Map<string, Function>
+
+  constructor (targetWindow: Window, targetUrl: string) {
+    this.targetWindow = targetWindow
+    this.targetUrl = (new URL(targetUrl)).origin // Use URL.origin to remove trailing '/' on targetUrl
+    this.handlers = new Map<string, Function>()
+  }
+
+  sendCommand = <T> (command: LedgerFrameCommand): Promise<T | LedgerErrorsCodes > => {
+    return new Promise<T | LedgerErrorsCodes>(async (resolve) => {
+      // Set handler for the response by passing the resolve function to be run
+      // when targetWindow responds using the same command.id.
+      // This allows us to simply `await` the sendCommand response
+      if (!this.addCommandHandler(command.id, resolve)) {
+        resolve(LedgerErrorsCodes.CommandInProgress)
+        return
+      }
+      this.targetWindow.postMessage(command, this.targetUrl)
+    })
+  }
+
+  protected addCommandHandler = (id: string, listener: Function): boolean => {
```

Comment:
The handler function should have two potential types.  The type is `(command: LedgerFrameCommand) => Promise<LedgerFrameResponse>` if it's a 'standard' handler that's invoked via a sendCommand call, e.g. handleUnlock.  In this case it takes the command as input, and returns the response.

Otherwise the handler is a resolve function that is called when a response message for an earlier sendCommand is received by the transport.  In that case the handler passes the received response message as input to the resolve function completing the cycle. In that case the type is `((response: LedgerFrameResponse) => void)`.

I have updated the types of addCommandHandler and the type of the handlers map initialized in the constructor to reflect this, however more work needs to be done to update the overall typing of the handlers attribute from `Map<string, Function>`.